cpp/src/community/k_truss_impl.cuh

```diff
@@ -41,7 +42,214 @@
 #include <thrust/transform.h>
 #include <thrust/tuple.h>
 
+#include <chrono>
+using namespace std::chrono;
+
 namespace cugraph {
+template <typename vertex_t, typename edge_t, bool multi_gpu>
+void order_edge_based_on_dodg(raft::handle_t const& handle,
+                              graph_view_t<vertex_t, edge_t, false, multi_gpu>& graph_view,
+                              raft::device_span<vertex_t> edgelist_srcs,
+                              raft::device_span<vertex_t> edgelist_dsts)
+
+{
```

Comment:
So, here, what we are actually doing is this.

For (src, dst) pairs in edgelist_srcs & edgelist_dsts, return (src, dst) if (src, dst) exists in graph_view, and return (dst, src) if not.

I think this code is overly complex.

If single-GPU, this code looks fine.

In multi-GPU, you first need to find unique (src, dst) pairs, shuffle, call has_edge, shuffle_back. Now you have (src, dst, exist) triplets. Sort them for binary search. Then, you can iterate over the input pairs and flip or not based on whether the edge exists or not.

In the current code, you are finding unique edges after shuffling. But better do this at the beginning to reduce computation/communication. And for temporary variables, try to minimize their scopes. Some variables' scopes are too large and this makes code harder to understand.